src/wp-admin/includes/theme.php

```diff
@@ -711,17 +711,20 @@ function wp_prepare_themes_for_js( $themes = null ) {
 		$is_block_theme         = $theme->is_block_theme();
 
 		if ( $is_block_theme && $can_edit_theme_options ) {
-			$customize_action = esc_url( admin_url( 'site-editor.php' ) );
+			$customize_action = admin_url( 'site-editor.php' );
+			if ( $current_theme !== $slug ) {
+				$customize_action = add_query_arg( 'theme_preview', $slug, $customize_action );
+			}
 		} elseif ( ! $is_block_theme && $can_customize && $can_edit_theme_options ) {
-			$customize_action = esc_url(
-				add_query_arg(
-					array(
-						'return' => urlencode( sanitize_url( remove_query_arg( wp_removable_query_args(), wp_unslash( $_SERVER['REQUEST_URI'] ) ) ) ),
-					),
-					wp_customize_url( $slug )
-				)
-			);
+			$customize_action = wp_customize_url( $slug );
 		}
+		$customize_action = add_query_arg(
+			array(
+				'return' => urlencode( sanitize_url( remove_query_arg( wp_removable_query_args(), wp_unslash( $_SERVER['REQUEST_URI'] ) ) ) ),
+			),
+			$customize_action
+		);
+		$customize_action = esc_url( $customize_action );
```

Comment:
Escaping should always happen at the time of output, see https://developer.wordpress.org/apis/security/escaping/#toc_3.
```suggestion
```

Reply:
Makes sense to me, when would you escape it then, [here](https://github.com/WordPress/wordpress-develop/blob/7cfc7a46f9d9182b3c99e985f5b28b36b284bb92/src/wp-admin/themes.php#L908)? Sorry I'm not familiar with the templating we are using here for this.

Reply:
After thinking about this and reviewing [the old cod](https://github.com/WordPress/wordpress-develop/blob/04c8d778558b1662e659c97169fcef47a25ac8e9/src/wp-admin/includes/theme.php#L714)e with @draganescu we think that it makes sense to escape here. The old code was already escaping there, and the variable is then passed as a larger JS object to the handlebar templating, so this is the last place for the code to be escaped in PHP

Reply:
Fair point, given the old code did that already, I guess it's not a big deal.